### PR TITLE
Implement RFC 1560 behind `#![feature(item_like_imports)]`

### PR DESCRIPTION
**Pull request by jseyfried**
This implements https://github.com/rust-lang/rfcs/pull/1560 (cc #35120) behind the `item_like_imports` feature gate.

The [RFC text](https://github.com/rust-lang/rfcs/blob/master/text/1560-name-resolution.md#changes-to-name-resolution-rules) describes the changes to name resolution enabled by `#![feature(item_like_imports)` in detail. To summarize,
- Items and named imports shadow glob imports.
- Multiple globs can import the same name if the name is unused or the imports are shadowed.
- Multiple globs can import the same name if the imports are of the same item (following re-exports).
  - The visibility of such a name is the maximum visibility of the imports.
  - Equivalently, adding a glob import will never reduce the visibility of a name, nor will removing one increase it.
- Non-prelude private imports can be used wherever we currently allow private items to be used.
  - Prelude-imported names are unaffected, i.e. they continue to be usable only in lexical scopes.
- Globs import all visible names, not just public names.
  - Equivalently, glob importing from an ancestor module imports all of the ancestor's names, and glob importing from other modules is unchanged.

r? @nrc

### CODE REVIEW

**Comment by jseyfried** [comment]
cc @petrochenkov

File: src/librustc_resolve/build_reduced_graph.rs [hunk 1]
```diff
@@ -390,7 +388,7 @@ impl<'b> Resolver<'b> {
                 debug!("(building reduced graph for external crate) building module {} {:?}",
                        name, vis);
                 let parent_link = ModuleParentLink(parent, name);
-                let module = self.new_module(parent_link, Some(def), true);
+                let module = self.new_module(parent_link, Some(def), DUMMY_NODE_ID);
```
Comment (nrc) [inline src/librustc_resolve/build_reduced_graph.rs]: what does using DUMMY_NODE_ID mean? Seems like it is just ignored when searching, so it would probably be better to use an Option<NodeId> and None here.

File: src/librustc_resolve/lib.rs [hunk 2]
```diff
@@ -753,6 +753,9 @@ pub struct ModuleS<'a> {
     parent_link: ParentLink<'a>,
     def: Option<Def>,
 
+    // The node id of the closest normal module (`mod`) ancestor (including this module).
```
Comment (nrc) [inline src/librustc_resolve/lib.rs]: So, for example for an item defined inside a function inside a module `foo`, this would be the id for `foo`? What is it used for? Evaluating `super` in paths, meaning of pub(restricted), is that all?

File: src/test/compile-fail/imports/shadowing.rs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+fn h() {
+    use foo::*;
+    use bar::*;
+    fn f() -> bool { true } // This also shadows the globs
+    let _: bool = f();
+}
+
+#[rustc_error]
+fn main() {} //~ ERROR compilation successful
```
Comment (nrc) [inline src/test/compile-fail/imports/shadowing.rs]: Could this be a run-pass test?

Comment (nrc) [inline src/test/compile-fail/imports/shadowing.rs]: Could you add to the run-pass situations where there appears to be shadowing but isn't because of namespaces, and where there is shadowing in one namespace, but not another.

File: src/librustc_resolve/resolve_imports.rs [hunk 4]
```diff
@@ -52,7 +52,10 @@ pub enum ImportDirectiveSubclass<'a> {
         value_result: Cell<Result<&'a NameBinding<'a>, Determinacy>>,
         type_result: Cell<Result<&'a NameBinding<'a>, Determinacy>>,
     },
-    GlobImport { is_prelude: bool },
+    GlobImport {
+        is_prelude: bool,
+        max_vis: Cell<ty::Visibility>, // The visibility of the greatest reexport.
```
Comment (nrc) [inline src/librustc_resolve/resolve_imports.rs]: Is this basically an optimisation to not even look at globs which would import nothing? Because otherwise, I would not expect the visibility to matter here.

File: src/librustc_resolve/resolve_imports.rs [hunk 5]
```diff
@@ -356,8 +356,11 @@ impl<'a> Resolver<'a> {
         };
 
         // Define `binding` in `module`s glob importers.
-        if binding.vis == ty::Visibility::Public {
-            for directive in module.glob_importers.borrow_mut().iter() {
+        for directive in module.glob_importers.borrow_mut().iter() {
+            if match self.new_import_semantics {
+                true => self.is_accessible_from(binding.vis, directive.parent),
+                false => binding.vis == ty::Visibility::Public,
+            } {
```
Comment (nrc) [inline src/librustc_resolve/resolve_imports.rs]: optional nit: I would pull the match out into a variable, I find this construction hard to read

File: src/test/compile-fail/imports/duplicate.rs [hunk 6]
```diff
@@ -33,6 +33,20 @@ mod e {
...
+    pub use a::*; //~ NOTE `foo` could resolve to the name imported here
+    pub use f::*; //~ NOTE `foo` could also resolve to the name imported here
+}
+
 fn main() {
     e::foo();
+    f::foo(); //~ ERROR `foo` is ambiguous
+              //~| NOTE Consider adding an explicit import of `foo` to disambiguate
+    g::foo(); //~ ERROR `foo` is ambiguous
+              //~| NOTE Consider adding an explicit import of `foo` to disambiguate
```
Comment (nrc) [inline src/test/compile-fail/imports/duplicate.rs]: Could we get a run-pass test too for where the ambiguity doesn't matter?

File: src/test/compile-fail/task-rng-isnt-sendable.rs [hunk 7]
```diff
@@ -10,10 +10,10 @@
 
 // ensure that the ThreadRng isn't/doesn't become accidentally sendable.
 
-use std::rand; //~ ERROR: module `rand` is private
+use std::__rand::ThreadRng;
```
Comment (nrc) [inline src/test/compile-fail/task-rng-isnt-sendable.rs]: Shouldn't we need the feature flag for this change to happen?

Comment (jseyfried) [inline src/test/compile-fail/task-rng-isnt-sendable.rs]: The change to this test is orthogonal to the rest of the PR -- c.f. https://github.com/rust-lang/rust/pull/35776#discussion_r75729787

File: src/test/compile-fail/imports/shadowing.rs (hunk 3 above)
Comment (jseyfried) [inline src/test/compile-fail/imports/shadowing.rs]: Will do.

File: src/test/compile-fail/imports/duplicate.rs (hunk 6 above)
Comment (jseyfried) [inline src/test/compile-fail/imports/duplicate.rs]: Will do.

File: src/librustc_resolve/build_reduced_graph.rs (hunk 1 above)
Comment (jseyfried) [inline src/librustc_resolve/build_reduced_graph.rs]: `DUMMY_NODE_ID` means that the module is from an extern crate, so its nearest normal ancestor doesn't have a node id. `normal_ancestor_id` is never used on modules from extern crates, so the the choice of node id isn't particularly relevant.

File: src/librustc_resolve/lib.rs (hunk 2 above)
Comment (jseyfried) [inline src/librustc_resolve/lib.rs]: > So, for example for an item defined inside a function inside a module foo, this would be the id for foo?

Yeah. It is used for evaluating `self` and `super` in paths, resolving `pub(restricted)` to a `ty::Visibility`, more efficient implementation of `NodeIdTree`, and (most importantly) checking the visibility of imports and paths in the module (c.f. [`is_accessible`](https://github.com/rust-lang/rust/pull/35894/files#diff-aeb0880081a991f34aef2ab889e1fb7aL3278)).

File: src/librustc_resolve/resolve_imports.rs (hunk 4 above)
Comment (jseyfried) [inline src/librustc_resolve/resolve_imports.rs]: No, this visibility doesn't matter during import resolution as you expect. It is only used in `finalize_import`s [here](https://github.com/rust-lang/rust/pull/35894/files#diff-9a6ea146fd5deda078e1d8402f5cf4eeR600) to enforce the requirement that a glob must import something with at least the glob's visibility.

It would also be sufficient to use `imports_name_with_sufficient_visibility: bool` instead of `max_vis: Cell<ty::Visibility>`, but I think knowing `max_vis` could be useful later for better error messages, for example;
```rust
mod foo {
    pub(crate) fn f() {}
}
pub use foo::*; //~ ERROR This glob must import something with `pub` visiblity
                //~^ NOTE consider using `pub(crate) use foo::*;`

Comment (jseyfried) [inline src/librustc_resolve/resolve_imports.rs]: (c.f. the last paragraph in [this section](https://github.com/rust-lang/rfcs/blob/master/text/1560-name-resolution.md#re-exports-namespaces-and-visibility) of the RFC text)

File: src/test/compile-fail/imports/duplicate.rs (hunk 6 above)
Comment (jseyfried) [inline src/test/compile-fail/imports/duplicate.rs]: Done.

File: src/test/compile-fail/imports/shadowing.rs (hunk 3 above)
Comment (jseyfried) [inline src/test/compile-fail/imports/shadowing.rs]: Done.

File: src/librustc_resolve/resolve_imports.rs (hunk 4 above)
Comment (nrc) [inline src/librustc_resolve/resolve_imports.rs]: Ok, that sounds fine, could you add a comment about this please?

File: src/librustc_resolve/build_reduced_graph.rs (hunk 1 above)
Comment (nrc) [inline src/librustc_resolve/build_reduced_graph.rs]: Then I think it would be better to use Option<NodeId> here, rather than DUMMY_NODE_ID, this seems like the moral equivalent of using null. Put another way, it would be nice to enforce "normal_ancestor_id is never used on modules from extern crates" statically.

**Comment by nrc** [comment]
r=me with the last comments addressed.

File: src/librustc_resolve/resolve_imports.rs (hunk 4 above)
Comment (jseyfried) [inline src/librustc_resolve/resolve_imports.rs]: Will do.

File: src/librustc_resolve/build_reduced_graph.rs (hunk 1 above)
Comment (jseyfried) [inline src/librustc_resolve/build_reduced_graph.rs]: Will do, but that would still be enforced dynamically (with `.unwrap()`) unless we separate local and external modules at the type level.

File: src/librustc_resolve/resolve_imports.rs (hunk 4 above)
Comment (jseyfried) [inline src/librustc_resolve/resolve_imports.rs]: Done.
